sentry/src/main/java/io/sentry/SentryOptions.java

```diff
@@ -1,60 +1,311 @@
+
+    private final Lookup lookup;
+    private final Dsn dsn;
+    private final SentryClientFactory clientFactory;
+
+    /**
+     * Creates a new instance of the options using the provided parameters.
+     *
+     * @param lookup        the lookup to locate the configuration with
+     * @param dsn           the dsn to use
+     * @param clientFactory the client factory to use
+     * @throws NullPointerException if any of the parameters is null
+     */
+    public SentryOptions(Lookup lookup, Dsn dsn, SentryClientFactory clientFactory) {
+        this.lookup = requireNonNull(lookup, "lookup");
+        this.dsn = requireNonNull(dsn, "dsn");
+        this.clientFactory = requireNonNull(clientFactory, "clientFactory");
+    }
+
+    /**
+     * Creates new options using the provided lookup instance. The DSN and the client factory are obtained using the
+     * provided lookup instance.
+     *
+     * @param lookup the lookup to locate the configuration with
+     * @return new instance of SentryOptions
+     */
+    public static SentryOptions from(Lookup lookup) {
+        return from(lookup, (String) null, null);
+    }
+
+    /**
+     * Creates new options using the provided lookup instance and optional DSN. If the provided DSN is null, the DSN
+     * to use is looked up using the provided lookup instance.
+     *
+     * @param lookup the lookup to locate the configuration with
+     * @param dsn    the optional dsn to use
+     * @return new instance of SentryOptions
+     */
+    public static SentryOptions from(Lookup lookup, @Nullable String dsn) {
+        return from(lookup, dsn, null);
+    }
+
+    /**
+     * Creates new options using the provided lookup instance and optional DSN and client factory. If the provided
+     * DSN or client factory is null, the DSN and client factory to use is looked up using the provided lookup instance.
+     *
+     * @param lookup  the lookup to locate the configuration with
+     * @param dsn     the optional dsn to use or null if the value should be located in the config
+     * @param factory the client factory to use or null if the value should be located in the config
+     * @return new instance of SentryOptions
+     */
+    public static SentryOptions from(Lookup lookup, @Nullable String dsn, @Nullable SentryClientFactory factory) {
+        return from(lookup, resolveDsn(lookup, dsn), factory);
+    }
+
+    private static SentryOptions from(Lookup lookup, Dsn dsn, @Nullable SentryClientFactory factory) {
+        if (factory == null) {
+            factory = SentryClientFactory.instantiateFrom(lookup, dsn);
+        }
+
+        if (factory == null) {
+            logger.error("Failed to find a Sentry client factory in the provided configuration. Will continue"
+                    + " with a dummy implementation that will send no data.");
+
+            factory = new InvalidSentryClientFactory();
```

Comment:
Can we call it NoOpSentryClientFactory or something like that? Invalid seems like I wouldn't want that instance around.

Reply:
I did this to emulate the behavior of `SentryClientFactory#getClient()` which may return `null` but as far as I looked in the codebase such possibility is nowhere taken into account and would result in NPEs.

Because `SentryOptions` doesn't return a client but rather a client factory, I needed to use a factory that would return null client under the same circumstances as the factory. I believe such factory is actually invalid, because it returns null client, which the callers don't expect, hence the naming.